sirepo/package_data/static/js/sirepo.js

```diff
@@ -1556,14 +1544,154 @@ SIREPO.app.service('validationService', function(utilities) {
+        catch (e) {
+        }
+        return null;
+    }
 
-SIREPO.app.factory('frameCache', function(appState, panelState, requestSender, authState, $interval, $rootScope, $timeout) {
+    function open() {
+        if (! window.indexedDB || ! SIREPO.authState.uiWebSocket) {
+            return;
+        }
+        const r = window.indexedDB.open('srCache', 1);
+        r.onsuccess = (event) => {
+            db = event.target.result;
+            removeOldRecords();
+        };
+        r.onupgradeneeded = (event) => {
+            db = event.target.result;
+            if (db.objectStoreNames.contains(STORE)) {
+                db.deleteObjectStore(STORE);
+            }
+            const o = db.createObjectStore(STORE);
+            o.createIndex('simId', '_srcache_simId', { unique: false });
+            o.createIndex('mru', '_srcache_mru', { unique: false });
+        };
+    }
+
+    function removeOldRecords() {
+        const o = objectStore('readonly');
+        if (! o) {
+            return;
+        }
+        const expired = [];
+        const d = new Date().getTime();
+        o.index('mru').openKeyCursor().onsuccess = (event) => {
+            const c = event.target.result;
+            if (c) {
+                if ((d - c.key) > EXPIRY_TIME) {
+                    expired.push(c.primaryKey);
+                }
+                c.continue();
+            }
+            else {
+                deleteKeys(expired);
+            }
+        };
+    }
+
+    self.clearFrames = (simId, modelName) => {
+        // deletes frames by simId, or (simId, modelName)
+        const o = objectStore('readonly');
+        if (o) {
+            const keys = [];
+            o.index('simId').openCursor(window.IDBKeyRange.only(simId)).onsuccess = (event) => {
+                const c = event.target.result;
+                if (c) {
+                    if ((! modelName) || modelName === c.value._srcache_modelName) {
+                        keys.push(c.primaryKey);
+                        c.continue();
+                    }
+                }
+                else {
+                    deleteKeys(keys);
+                }
+            };
+        }
+    };
+
+    self.getFrame = (frameId, modelName, callback) => {
+        const o = objectStore('readonly');
+        if (! o) {
+            invokeCallback(callback, null);
+            return;
+        }
+        const c = o.get(key(FRAME, frameId));
+        c.onsuccess = (event) => {
+            const d = event.target.result;
+            invokeCallback(callback, d);
+            if (d) {
+                // update MRU (most recently used)
+                self.saveFrame(frameId, modelName, d);
+            }
+        };
+        c.onerror = () => {
+            invokeCallback(callback, null);
+        };
+    };
+
+    self.saveFrame = (frameId, modelName, data) => {
+        const o = objectStore('readwrite');
+        if (o) {
+            data._srcache_mru = new Date().getTime();
```

Comment:
I think "mru" should be updateTime, which is a term used elsewhere in the code. The index is updateTime as well.  It's not actually MRU, and instead it has a fixed cutoff.

Maybe over optimization, but creating a new Date every time seems unnecessary.